pkg/aws/services/tagging.go

```diff
@@ -0,0 +1,97 @@
+package services
+
+import (
+	"context"
+	"github.com/aws/aws-application-networking-k8s/pkg/utils"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	taggingapi "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi"
+	taggingapiiface "github.com/aws/aws-sdk-go/service/resourcegroupstaggingapi/resourcegroupstaggingapiiface"
+)
+
+//go:generate mockgen -destination tagging_mocks.go -package services github.com/aws/aws-application-networking-k8s/pkg/aws/services Tagging
+
+const (
+	resourceTypePrefix = "vpc-lattice:"
+
+	ResourceTypeTargetGroup = resourceTypePrefix + "targetgroup"
+
+	maxArnsPerGetResourcesApi = 100
+)
+
+type Tags = map[string]*string
+
+type Tagging interface {
+	taggingapiiface.ResourceGroupsTaggingAPIAPI
+
+	// Receives a list of arns and returns arn-to-tags map.
+	GetTagsFromArns(ctx context.Context, arns []*string) (map[string]Tags, error)
+
+	// Finds one resource that matches the given set of tags.
+	FindResourceWithTags(ctx context.Context, resourceType string, tags Tags) (*string, error)
+}
+
+type defaultTagging struct {
+	taggingapiiface.ResourceGroupsTaggingAPIAPI
+}
+
+func (t *defaultTagging) GetTagsFromArns(ctx context.Context, arns []*string) (map[string]Tags, error) {
+	chunks := utils.Chunks(arns, maxArnsPerGetResourcesApi)
+	result := make(map[string]Tags)
+
+	for _, chunk := range chunks {
+		input := &taggingapi.GetResourcesInput{
+			ResourceARNList: chunk,
+		}
+		err := t.GetResourcesPagesWithContext(ctx, input, func(page *taggingapi.GetResourcesOutput, lastPage bool) bool {
+			for _, r := range page.ResourceTagMappingList {
+				result[*r.ResourceARN] = convertTags(r.Tags)
+			}
+			return true
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+func (t *defaultTagging) FindResourceWithTags(ctx context.Context, resourceType string, tags Tags) (*string, error) {
+	input := &taggingapi.GetResourcesInput{
+		TagFilters:          convertTagsToFilter(tags),
+		ResourceTypeFilters: []*string{aws.String(resourceType)},
+	}
+	resp, err := t.GetResourcesWithContext(ctx, input)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.ResourceTagMappingList) == 0 {
+		return nil, NewNotFoundError("tag", "matching criteria")
+	}
+	// assume one result
+	return resp.ResourceTagMappingList[0].ResourceARN, nil
```

Comment:
I think assumption about how many results to expect should be on caller of this function, not here. So returning a slice of results would be better. Also empty slice as return an easy check for not found.

Reply:
makes sense